ownCloud/Client/Actions/Action.swift

```diff
@@ -0,0 +1,218 @@
+	class var features : [String : Any]? { return nil }
+
+	// MARK: - Extension creation
+	class var actionExtension : ActionExtension {
+		let objectProvider : OCExtensionObjectProvider = { (_ rawExtension, _ context, _ error) -> Any? in
+			if let actionExtension = rawExtension as? ActionExtension,
+				let actionContext   = context as? ActionContext {
+				return self.init(for: actionExtension, with: actionContext)
+			}
+
+			return nil
+		}
+
+		let customMatcher : OCExtensionCustomContextMatcher  = { (context, priority) -> OCExtensionPriority in
+
+			guard let actionContext = context as? ActionContext else {
+				return priority
+			}
+
+			if self.applicablePosition(forContext: actionContext) == .none {
+				// Exclude actions whose applicablePosition returns .none
+				return .noMatch
+			}
+
+			if actionContext.items[0].type == OCItemType.collection, identifier!.rawValue == "com.owncloud.action.openin" {
+				return .noMatch
+			}
+
+			let priority = OCExtensionPriority(rawValue: priority.rawValue + UInt(self.applicablePosition(forContext:actionContext).rawValue))!
+			return priority
+
+			// Additional filtering (f.ex. via OCClassSettings, Settings) goes here
+		}
+
+		return ActionExtension(name: name!, category: category!, identifier: identifier!, locations: locations, features: features, objectProvider: objectProvider, customMatcher: customMatcher)
+	}
+
+	// MARK: - Extension matching
+	class func applicablePosition(forContext: ActionContext) -> ActionPosition {
+		return .middle
+	}
+
+	// MARK: - Finding actions
+	class func sortedApplicableActions(for context: ActionContext) -> [Action] {
+		var sortedActions : [Action] = []
+
+		if let matches = try? OCExtensionManager.shared.provideExtensions(for: context) {
+			for match in matches {
+				if let action = match.extension.provideObject(for: context) as? Action {
+					sortedActions.append(action)
+				}
+			}
+		}
+
+		sortedActions.sort { (action1, action2) -> Bool in
+			return action1.position.rawValue < action2.position.rawValue
+		}
+
+		return sortedActions
+	}
+
+	// MARK: - Action metadata
+	var context : ActionContext
+	var actionExtension: ActionExtension
+	var core : OCCore
+
+	// MARK: - Action creation
+	required init(for actionExtension: ActionExtension, with context: ActionContext) {
+		self.actionExtension = actionExtension
+		self.context = context
+		self.core = context.core!
+
+		super.init()
+	}
+
+	// MARK: - Execution metadata
+	var progressHandler : ActionProgressHandler?     // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+	var completionHandler : ActionCompletionHandler? // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+	var beforeRunHandler: ActionBeforeRunHandler? // to be filled before calling run(), provideStaticRow(), provideContextualAction(), etc. if desired
+
+	// MARK: - Action implementation
+	func run() {
+		if beforeRunHandler != nil {
```

Comment:
Might be a good idea to move the code running `beforeRunHandler` to a separate `willRun` method, so that
- subclasses can just call `self.willRun()`
- more code can be added before without having to change the subclasses